apps/prairielearn/src/ee/lib/billing/plans.ts

```diff
@@ -0,0 +1,154 @@
+import { z } from 'zod';
+import { loadSqlEquiv, queryRow, queryRows, runInTransactionAsync } from '@prairielearn/postgres';
+import {
+  EnumPlanGrantType,
+  Institution,
+  InstitutionSchema,
+  PlanGrant,
+  PlanGrantSchema,
+} from '../../../lib/db-types';
+import { PLAN_NAMES, PlanName } from './plans-types';
+import { insertPlanGrant, updatePlanGrant, deletePlanGrant } from '../../models/plan-grants';
+import {
+  insertCourseInstanceRequiredPlan,
+  deleteCourseInstanceRequiredPlan,
+} from '../../models/course-instance-required-plans';
+
+const sql = loadSqlEquiv(__filename);
+
+export interface DesiredPlan {
+  plan: PlanName;
+  grantType: EnumPlanGrantType;
+}
+
+type InstitutionPlanGrantContext = Pick<PlanGrant, 'institution_id'>;
+type CourseInstancePlanGrantContext = Pick<PlanGrant, 'institution_id' | 'course_instance_id'>;
+type EnrollmentPlanGrantContext = Pick<
+  PlanGrant,
+  'institution_id' | 'course_instance_id' | 'enrollment_id'
+>;
+type UserPlanGrantContext = Pick<PlanGrant, 'user_id'>;
+type PlanGrantContext =
+  | InstitutionPlanGrantContext
+  | CourseInstancePlanGrantContext
+  | EnrollmentPlanGrantContext
+  | UserPlanGrantContext;
+
+export async function getPlanGrantsForInstitution(institution_id: string): Promise<PlanGrant[]> {
+  return queryRows(sql.select_plan_grants_for_institution, { institution_id }, PlanGrantSchema);
+}
+
+export async function getPlanGrantsForCourseInstance(
+  course_instance_id: string,
+): Promise<PlanGrant[]> {
+  return queryRows(
+    sql.select_plan_grants_for_course_instance,
+    { course_instance_id },
+    PlanGrantSchema,
+  );
+}
+
+export async function getRequiredPlansForCourseInstance(
+  course_instance_id: string,
+): Promise<PlanName[]> {
+  return queryRows(
+    sql.select_required_plans_for_course_instance,
+    { course_instance_id },
+    z.enum(PLAN_NAMES),
+  );
+}
+
+export async function updateRequiredPlansForCourseInstance(
+  course_instance_id: string,
+  plans: PlanName[],
+) {
+  await runInTransactionAsync(async () => {
+    // TODO: locking?
```

Comment:
I suppose so, although it doesn't seem very likely to be important in practice. We could always take a row lock on the CI.

Reply:
Yeah, that's probably true. I think we can go without locking here.

Reply:
Comments removed in 8a94eff.